File: code/computational_geometry/src/axis_aligned_bounding_box_collision/axis_aligned_bounding_box_collision.cpp
#include <iostream>
// Part of Cosmos by OpenGenus Foundation
struct Vector
{
    int x;
    int y;
};

struct Shape
{
    Vector center;
    int width;
    int height;
};

bool checkAABBCollision(Shape &a, Shape &b)
{
    // change '<' to '<=' if you want to include edge touching as a collision
    return (abs(a.center.x - b.center.x) * 2 < (a.width + b.width)) &&
           (abs(a.center.y - b.center.y) * 2 < (a.height + b.height));
}

int main()
{
    Shape a = { Vector {3, 3}, 4, 5 };
    Shape b = { Vector {9, 3}, 6, 4 };

    // 0 - no collision
    std::cout << checkAABBCollision(a, b) << std::endl;

    a = { Vector {3, 3}, 4, 5 };
    b = { Vector {7, 3}, 6, 4 };

    // 1 - collision
    std::cout << checkAABBCollision(a, b) << std::endl;

    a = { Vector {3, 10}, 4, 6 };
    b = { Vector {3, 5}, 6, 6 };

    // 1 - collision
    std::cout << checkAABBCollision(a, b) << std::endl;
}

File: cosmos-master/code/computational_geometry/axis_aligned_bounding_box_collision/axis_aligned_bounding_box_collision.go
package main
// Part of Cosmos by OpenGenus Foundation

import (
	"fmt"
	"math"
)

type vector struct {
	x, y float64
}

type shape struct {
	center vector
	height, width float64
}

func checkAABBCollision(a, b shape) bool {
	// change '<' to '<=' if you want to include edge touching as a collision
	return (math.Abs(a.center.x - b.center.x) *2 < (a.width + b.width)) &&
		(math.Abs(a.center.y - b.center.y) * 2 < (a.height + b.height))
}

func main() {
	shapeA := shape{center: vector{3, 3}, height: 5, width: 4}
	shapeB := shape{center: vector{9, 3}, height: 4, width: 6}

	// 0 - no collision
	fmt.Println(checkAABBCollision(shapeA, shapeB))

	shapeA = shape{center: vector{3, 3}, height: 5, width: 4}
	shapeB = shape{center: vector{7, 3}, height: 4, width: 6}

	// 1 - collision
	fmt.Println(checkAABBCollision(shapeA, shapeB))

	shapeA = shape{center: vector{3, 10}, height: 6, width: 4}
	shapeB = shape{center: vector{3, 5}, height: 6, width: 6}

	// 1 - collision
	fmt.Println(checkAABBCollision(shapeA, shapeB))
}
